### PR TITLE
Book entity, repository, and service tiers have been added.

### PR DESCRIPTION
**Pull request by oleksandr-chernetskyi**

### CODE REVIEW

File: src/main/java/onlinebookstore/model/Book.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Entity
+@Data
```
Comment (Rommelua) [inline src/main/java/onlinebookstore/model/Book.java:11]: Better do not use @Data in @Entity class

File: src/main/java/onlinebookstore/model/Book.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+public class Book {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private String title;
+    private String author;
+    private String isbn;
+    private BigDecimal price;
+    private String description;
+    private String coverImage;
```
Comment (Rommelua) [inline src/main/java/onlinebookstore/model/Book.java:21]: add constraints

**Review by Rommelua — CHANGES_REQUESTED** [review]

File: src/main/java/onlinebookstore/model/Book.java (hunk 2 above)
Comment (sokimaaa) [reply src/main/java/onlinebookstore/model/Book.java]: not fixed

File: src/main/java/onlinebookstore/model/Book.java (hunk 1 above)
Comment (sokimaaa) [reply src/main/java/onlinebookstore/model/Book.java]: not fixed

File: src/main/java/onlinebookstore/model/Book.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import java.math.BigDecimal;
+import java.util.Objects;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Entity
+@Table(name = "books")
```
Comment (Rommelua) [inline src/main/java/onlinebookstore/model/Book.java:15]:
```suggestion
@Entity
@Getter
@Setter
@NoArgsConstructor
@Table(name = "books")
```

File: src/main/java/onlinebookstore/model/Book.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        this.coverImage = coverImage;
+    }
+
+    public boolean isDeleted() {
+        return isDeleted;
+    }
+
+    public void setDeleted(boolean deleted) {
+        isDeleted = deleted;
+    }
```
Comment (Rommelua) [inline src/main/java/onlinebookstore/model/Book.java:97]: you can delete get/set

File: src/main/java/onlinebookstore/model/Book.java [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        }
+        Book book = (Book) o;
+        return isDeleted == book.isDeleted
+                && Objects.equals(id, book.id)
+                && Objects.equals(title, book.title)
+                && Objects.equals(author, book.author)
+                && Objects.equals(isbn, book.isbn)
+                && Objects.equals(price, book.price)
+                && Objects.equals(description, book.description)
+                && Objects.equals(coverImage, book.coverImage);
```
Comment (Rommelua) [inline src/main/java/onlinebookstore/model/Book.java:115]: No need to check all fields - we have a unique business identifier for book
```suggestion
        return  Objects.equals(isbn, book.isbn);
```

File: src/main/java/onlinebookstore/model/Book.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, title, author, isbn, price, description, coverImage, isDeleted);
+    }
+
+    @Override
+    public String toString() {
```
Comment (Rommelua) [inline src/main/java/onlinebookstore/model/Book.java:124]: do you need this method?

File: src/main/java/onlinebookstore/repository/book/specification/DescriptionSpecificationProvider.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
+package onlinebookstore.repository.book.specification;
+
+import java.util.Arrays;
+import onlinebookstore.model.Book;
+import onlinebookstore.repository.SpecificationProvider;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DescriptionSpecificationProvider implements SpecificationProvider<Book> {
```
Comment (Rommelua) [inline src/main/java/onlinebookstore/repository/book/specification/DescriptionSpecificationProvider.java:10]: delete this class. Is it possible to search by full Description?

File: src/main/java/onlinebookstore/repository/book/specification/PriceSpecificationProvider.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+public class PriceSpecificationProvider implements SpecificationProvider<Book> {
+    private static final String PRICE = "price";
+
+    @Override
+    public Specification<Book> getSpecification(String[] parameters) {
+        return ((root, query, criteriaBuilder) -> root
+                .get(PRICE).in(Arrays.stream(parameters).toArray()));
```
Comment (Rommelua) [inline src/main/java/onlinebookstore/repository/book/specification/PriceSpecificationProvider.java:16]: price IN? It is not possible to use. You should implement a search between lower and higher boundaries for price or delete it at all

File: src/main/java/onlinebookstore/repository/book/specification/TitleSpecificationProvider.java [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    private static final String TITLE = "title";
+
+    @Override
+    public Specification<Book> getSpecification(String[] parameters) {
+        return ((root, query, criteriaBuilder) -> root
+                .get(TITLE).in(Arrays.stream(parameters).toArray()));
```
Comment (Rommelua) [inline src/main/java/onlinebookstore/repository/book/specification/TitleSpecificationProvider.java:16]: better have 'contains' logic for the title.
When you search "Harry Potter" you want to see all the parts. But here you will see nothing.

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Review by aaananas — APPROVED** [review]

**Review by sokimaaa — CHANGES_REQUESTED** [review]

**Review by Rommelua — CHANGES_REQUESTED** [review]
image
Only code connected with this task should be in the PR. Previous tasks should already be in main branch.

**Pull request merged**